Tasks/DeployVisualStudioTestAgent/TestAgentInstall.ps1

```diff
@@ -3,13 +3,24 @@ function Install-Product($SetupPath, $UserName, $Password, $ProductVersion, $Arg
 	$InstalledCheckRegKey = ("SOFTWARE\Microsoft\DevDiv\vstf\Servicing\{0}\testagentcore" -f $ProductVersion)
 	$InstalledCheckRegValueName = "Install"
 	$InstalledCheckRegValueData = "1"
-
-	$isProductExists = Get-ProductEntry -InstalledCheckRegKey $InstalledCheckRegKey -InstalledCheckRegValueName $InstalledCheckRegValueName -InstalledCheckRegValueData $InstalledCheckRegValueData
-        $testAgentFileExists = Test-Path "$env:SystemDrive\TestAgent\testagent"
-
-	if($testAgentFileExists -and $isProductExists)
+                
+	$isProductExists = Get-ProductEntry -InstalledCheckRegKey $InstalledCheckRegKey -InstalledCheckRegValueName $InstalledCheckRegValueName         
+
+        $versionToInstall = ((Get-Item $SetupPath).VersionInfo.FileVersion) 
+        $versionInstalled = (Get-ProductEntry -InstalledCheckRegKey $InstalledCheckRegKey -InstalledCheckRegValueName "version")
+
+        if($versionToInstall -ne $null)
+        {
+		$versionToInstall = $versionToInstall.split('.')
+        }
+
+        if($versionInstalled -ne $null)
+        {
+		$versionInstalled = $versionInstalled.split('.')
+        }
+        
+	if(($isProductExists -eq $InstalledCheckRegValueData) -and ($versionToInstall -ne $null) -and ($versionInstalled -ne $null) -and ($versionToInstall[0] -le $versionInstalled[0]) -and ($versionToInstall[1] -le $versionInstalled[1]) -and ($versionToInstall[2] -le $versionInstalled[2]))
```

Comment:
this condition should be simplified. Doesn't FileVersion class has any compare method?